Upgrade prow setup (#182)
metal3-io/project-infra: prow/README.md

# Prow

> Metal3 Prow dashboard: https://prow.apps.test.metal3.io

Reviewer: why `test`? why not `ci`?

Author: this is a new domain, previously it was http://prow.apps.ci.metal3.io, because Russell had DNS records pointing to his cluster IPs. We created a new record (`prow.apps.test.metal3.io` -> our IP)  with under the metal3.io domain to point to the new cluster.

  to manage access to all or part of a repo with this file.

* Tests will run automatically for PRs authored by **public** members of the
  `metal3-io` github organization.  Members of the github org can run
  `/ok-to-test` for PRs authored by those not in the github org.

See the [Prow command help](https://prow.apps.test.metal3.io/command-help) for
more information about who can run each prow command.

## GCS bucket

Google Cloud Storage (GCS) is required to store job artifacts. Follow the docs to set
up GCS and create a bucket with enough permissions.

1. Create a [bucket](https://cloud.google.com/storage/docs/creating-buckets)
1. Create a [service account](https://cloud.google.com/iam/docs/creating-managing-service-accounts)
1. On the created bucket, grant `storage.objects.create` access for the service account and `storage.objectViewer`
   permission to `allUser`.

## Setup

Prow was set up by following these instructions: https://github.com/kubernetes/test-infra/blob/master/prow/getting_started_deploy.md

1. Create DNS record(s) to point to your cluster.

1. Prow components are running on `default` namespace, while `test-pods` namespace

>    is used by pods running actual CI jobs.

Reviewer 2: ```suggestion
   is used by pods running the actual CI jobs.
```

1. [ghProxy](https://github.com/kubernetes/test-infra/tree/master/ghproxy) is a reverse proxy,
   and is meant to reduce API token usage. We need persistent volume for ghProxy, and for that
   purpose we rely on `hostPath` type of `PersistentVolume` in our cluster due to the limitation
   in the underlying infrastructure. Ideally, you should use a dynamic storage provisioner.

   Create a /mnt/prow directory to store the persistent volume data:

     ```shell
     $ sudo mkdir /mnt/prow
     ```

     Create the PersistentVolume:

     ```shell
     $ kubectl apply -f persistent_volume.yaml
     ```

1. Grant a user `cluster-admin` role in all namespaces to create cluster resources.

>     ```shell
>     $ kubectl create clusterrolebinding cluster-admin-binding-"${USER}" \
>       --clusterrole=cluster-admin --user="${USER}"
>     ```

Reviewer 2: ```suggestion
    ```shell
    USER=cluster-admin
    $ kubectl create clusterrolebinding cluster-admin-binding-"${USER}" \
      --clusterrole=cluster-admin --user="${USER}"
    ```
```

Author: I think this is correct, because we grant `--clusterrole=cluster-admin` permission to  ${USER}, any ${USER} can be different depending on where the cluster is running.

Reviewer 2: I think that user refers to the `service account`. 
If the OS user and service account are the same, then it is ok.

Author: I followed instructions here https://github.com/kubernetes/test-infra/blob/master/prow/getting_started_deploy.md#create-cluster-role-bindings.

> 1. Create hmac token to give to GitHub for webhook validation and generate a secret out of it.

Reviewer 3: Maybe:
```suggestion
1. Create hmac token to manage webhook validation and generate a secret out of it.
```

    ```shell
    $ openssl rand -hex 20 > /path/hmac-token
    $ kubectl create secret generic hmac-token --from-file=hmac=/path/hmac-token
    ```

1. Create a personal access token for the GitHub
   bot account. This should be done from [metal3-io-bot](https://github.com/metal3-io-bot)
   GitHub bot account. You can follow this [link](https://github.com/settings/tokens)

>    to create the token. When generating the token, ensure to have the following scopes checked in.

Reviewer 2: ```suggestion
   to create the token. When generating the token, make sure you have the following scopes checked in.
```

   - `public_repo` and `repo:status`
   - `repo scope` for private repos
   - `admin:org_hook` for a github org

    Create a secret out of that token:
    ```shell

>     $ kubectl create secret generic github-token --from-file=token=/path/github-token

Reviewer: does it have to be in any specific namespace ?

Author: same as below, no

    ```

1. Create a secret with the GCS credentials

   ```shell

>    $ kubectl create secret generic gcs-credentials --from-file=service-account.json=service-account.json

Reviewer: does it have to be in any specific namespace ?

Author: no, currently all the deployments are in default namespace. We can in the future move the to a separate namespace or even a dedicated cluster.

   ```

> 1. While applying prow manifests, we are going to create an Ingress object to redirect traffic to deck and hook services.

Reviewer 2: ```suggestion
1. While applying prow manifests, we are going to create an Ingress object to redirect traffic to the deck and hook services.
```

   We are installing [NGINX Ingress controller](https://kubernetes.github.io/ingress-nginx/deploy/) with helm, but you can
   install your preferred one.

   ```shell
   helm repo add ingress-nginx https://kubernetes.github.io/ingress-nginx
   helm install ingress-nginx/ingress-nginx --set controller.hostNetwork=true,controller.service.type="",controller.kind=DaemonSet --generate-name
   helm repo update
   ```

1. Apply Prow manifests

   ```shell
   kubectl apply -f manifests/*
   ```

>    This will create a couple of Deployments, Services, ServiceAccounts, Roles, RoleBindings and Ingress.

Reviewer 2: ```suggestion
   This will create a couple of Deployments, Services, ServiceAccounts, Roles, RoleBindings and Ingress resources.
```

## Secure Ingress with Cert-Manager and Let's Encrypt

1. Install [cert-manager](https://github.com/jetstack/cert-manager)

   ```
   $ helm repo add jetstack https://charts.jetstack.io
   $ helm repo update
   $ helm install \
   cert-manager jetstack/cert-manager \
   --namespace cert-manager \
   --create-namespace \
   --version v1.3.1 \
   --set installCRDs=true
   ```

   Verify that you have the following pods running
   ```shell
   $ kubectl get pods --namespace cert-manager

     NAME                                       READY   STATUS    RESTARTS   AGE
     cert-manager-5c6866597-zw7kh               1/1     Running   0          2m
     cert-manager-cainjector-577f6d9fd7-tr77l   1/1     Running   0          2m
     cert-manager-webhook-787858fcdb-nlzsq      1/1     Running   0          2m
   ```

> 1. Create a secret containing your Cloudflare API token

Reviewer 2: A sentence or two before this that explains the need for Cloudflare

Author: Yep, done.

   ```shell

>    $ kubectl create secret generic cloudflare-api-token-secret --from-literal "apikey=<API_KEY>" --namespace=cert-manager

Reviewer 3: Would be nice to have more information why do we need `cloudflare-api-token-secret` and the role of it in the prow setup.

Author: Added some notes.

   permission to `allUser`.

## Setup

Prow was set up by following these instructions: https://github.com/kubernetes/test-infra/blob/master/prow/getting_started_deploy.md

1. Create DNS record(s) to point to your cluster.

1. Prow components are running on `default` namespace, while `test-pods` namespace
   is used by pods running the actual CI jobs.

1. [ghProxy](https://github.com/kubernetes/test-infra/tree/master/ghproxy) is a reverse proxy,
   and is meant to reduce API token usage. We need persistent volume for ghProxy, and for that
   purpose we rely on `hostPath` type of `PersistentVolume` in our cluster due to the limitation
   in the underlying infrastructure. Ideally, you should use a dynamic storage provisioner.

   Create a /mnt/prow directory to store the persistent volume data:

     ```shell
     $ sudo mkdir /mnt/prow
     ```

     Create the PersistentVolume:

     ```shell
     $ kubectl apply -f persistent_volume.yaml
     ```

1. Grant a user `cluster-admin` role in all namespaces to create cluster resources.

    ```shell
    $ kubectl create clusterrolebinding cluster-admin-binding-"${USER}" \
      --clusterrole=cluster-admin --user="${USER}"
    ```

> 1. Create hmac token to for webhook validation and generate a secret out of it. Don't forget to add this token

Reviewer 3: ```suggestion
1. Create hmac token for webhook validation and generate a secret out of it. Don't forget to add this token
```
